### ISSUE TITLE
How about adding update!(opt, x::TrackedArray)?

### ISSUE THREAD

**Issue by kota7** [FluxML/Flux.jl#638]
Currently `update!` for optimizer is only defined (as far as I know) here:

https://github.com/FluxML/Flux.jl/blob/8386a49bf9a2950d51889b8af1ace178bf5169fa/src/optimise/train.jl#L5-L7

Here, `x` is parameter (or its data) and `\bar{x}` is gradient.  Since users often want to update the parameters to the negative gradient direction, how about adding something like this?

```julia
function update!(opt, x::TrackedArray)
    update!(x, -apply!(opt, x.data, x.grad))
end
```

Will this be useful?  I think this would be a direct counter part of `pytorch`'s `optimiser.step()`.

**Comment by thebhatman** [comment]
@kota7  the update! being called inside the function is taken from https://github.com/FluxML/Flux.jl/blob/master/src/tracker/lib/array.jl. Here update! is defined for x:: TrackedArray, like the way you suggested. So it is already existing.

**Comment by kota7** [comment]
@thebhatman Thank you for the reply.
I find this one

https://github.com/FluxML/Flux.jl/blob/d6cf116a740c2879ff04de95174d1528b7e43cd6/src/tracker/lib/array.jl#L68-L72

where users give how much you change the parameter as the second argument `Δ`.

My suggestion is to define `update!` for combination of an optimiser and parameter, i.e. `update!(optimiser, x :: TrackedArray)`, which essentially calculates `Δ` appropriately using the optimiser.  This way, users do not need to handle gradients directly.   I thought this would be useful since parameters are often updated in this manner.

**Comment by MikeInnes** [comment]
I think this was done at some point; the definition above was just a mistake.
